exercise/templates/exercise/exercise.html

```diff
@@ -79,28 +79,54 @@
 <script src="{% static 'exercise/duplicate_check.js' %}"></script>
 <script src="{% static 'exercise/chapter.js' %}"></script>
 <script
-    src="{% static 'exercise/language_link.js' %}"
-    data-set-lang="{{ instance|url:'set-enrollment-language' }}"></script>
+	src="{% static 'exercise/language_link.js' %}"
+	data-set-lang="{{ instance|url:'set-enrollment-language' }}"></script>
 <script>
 // Visualize loading in case of slow POST operations.
 /*$(function() {
-    var forms = $("#exercise-page-content").find("form:not([data-aplus-exercise] form)");
-    forms.on("submit", function(event) {
-        $("#submit-progress").removeClass("hide");
-    });
+	var forms = $("#exercise-page-content").find("form:not([data-aplus-exercise] form)");
+	forms.on("submit", function(event) {
+		$("#submit-progress").removeClass("hide");
+	});
 });*/
 
 // Add an Ajax exercise event listener to refresh the info column.
 window.addEventListener("message", function (event) {
-    if (event.data.type === "a-plus-refresh-stats") {
-        $("#submit-progress").addClass("hide");
-        var $stats = $("#exercise-info"),
-            url = $stats.data("url");
-        $stats.load(url, function() {
-            $stats.find('[data-toggle="tooltip"]').tooltip();
-        });
-    }
+	if (event.data.type === "a-plus-refresh-stats") {
+		$("#submit-progress").addClass("hide");
+		var $stats = $("#exercise-info"),
+			url = $stats.data("url");
+		$stats.load(url, function() {
+			$stats.find('[data-toggle="tooltip"]').tooltip();
+		});
+	}
 });
 </script>
+
+<script>
+var hash = window.location.hash
+
+//only register the events related to anchor links when theres a hash (e.g. #chapter-exercise-3)
+if (hash) {
+	//it isn't enough to just register a scroll event which disables setting the scroll
+	//location to the hash id. The scroll event fires every time the page loads, and
+	//whenever an exercise loads (as this changes the height of the page). To identify
+	//if a user manually scrolls we should receive 3 consecutive scroll events with no
+	//ajaxComplete events in between. An ajaxComplate event occuring signifies that the
+	//page scroll was caused by an exercise loading, and not the user.
+	var consecutiveScrollEvents = 0
+
+	window.addEventListener('scroll', () => {
+		consecutiveScrollEvents += 1
+	})
+
+	$( document ).ajaxComplete(() => {
```

Comment:
Did you consider adding an event handler like this to chapter.js? In chapter.js, you could identify which exercise is targeted by the `window.location.hash` and only react when that exercise is loaded. Though, when the exercises listed before the targeted exercise are loaded, that already changes the window's scroll position. Chapter.js loads the exercises in order from the first to the last.

Reply:
If by the time exercise 4 loads it's guaranteed that exercise 1, 2 and 3 have already been loaded then you could remove the ajax and scroll listener when the exercise referenced by the URL has loaded, as the page won't be scrolled further by loading exercises.

Reply:
It doesn't seem absolutely necessary to move the code to chapter.js. The code is related to what is done in chapter.js, though. Inside `exercise.html`, the JavaScript affects also full-view exercise pages outside chapters, but I don't think it is harmful. Those exercise pages usually don't use any hashes/anchors in the URL (the `#` part in the end of the URL).